registry/storage/blobwriter.go

```diff
@@ -379,28 +422,98 @@ func (bw *blobWriter) removeResources(ctx context.Context) error {
 	return nil
 }
 
-func (bw *blobWriter) Reader() (io.ReadCloser, error) {
-	// todo(richardscothern): Change to exponential backoff, i=0.5, e=2, n=4
-	try := 1
-	for try <= 5 {
-		_, err := bw.driver.Stat(bw.ctx, bw.path)
-		if err == nil {
-			break
+type blobWriterReader struct {
+	path       string
+	events     chan fsnotify.Event
+	blobWriter *blobWriter
+	driver     storagedriver.StorageDriver
+	ctx        context.Context
+	reader     io.ReadCloser
+	watcher    io.Closer
+	finished   chan struct{}
+}
+
+func (reader *blobWriterReader) Read(buff []byte) (int, error) {
+	reader.blobWriter.mutex.Lock()
+	inProgress := reader.blobWriter.IsInProgress()
+	defer reader.blobWriter.mutex.Unlock()
+	for true {
+		reader.blobWriter.mutex.Unlock()
+		count, err := reader.reader.Read(buff)
+		reader.blobWriter.mutex.Lock()
+		if err != nil && err != io.EOF {
+			return count, err
 		}
-		switch err.(type) {
-		case storagedriver.PathNotFoundError:
-			dcontext.GetLogger(bw.ctx).Debugf("Nothing found on try %d, sleeping...", try)
-			time.Sleep(1 * time.Second)
-			try++
-		default:
-			return nil, err
+		if count != 0 {
+			return count, nil
+		}
+		if err == io.EOF && !inProgress {
+			if reader.blobWriter.lastError != nil {
+				err = reader.blobWriter.lastError
+			}
+			return count, err
 		}
+		// At this point we have returned all the data we have available, we need to wait for
+		// further data to arrive or the download to be interrupted.
+		reader.blobWriter.mutex.Unlock()
+		select {
+		case <-reader.events:
+		case <-reader.finished:
+		case <-time.After(60 * time.Second):
+			logrus.Debugf("Timed out waiting for events ...")
+		}
+		reader.blobWriter.mutex.Lock()
 	}
+	return 0, io.EOF
+
+}
 
-	readCloser, err := bw.driver.Reader(bw.ctx, bw.path, 0)
+func (reader *blobWriterReader) Close() error {
+	return reader.blobWriter.ReleaseResources()
+}
+
+var _ io.ReadCloser = &blobWriterReader{}
+var _ ReadableWriter = &blobWriter{}
```

Comment:
Why are these two created?

Reply:
This is a way to enforce that a struct implements all the methods on the interface. Sort of the "implements" keyword in duck typing as I understood. :)

Reply:
Ah, I've seen this one in [effective go](https://golang.org/doc/effective_go#blank_implements): 

```golang
var _ json.Marshaler = (*RawMessage)(nil)
```

but it looked slightly different (casting instead of creating?).

Reply:
That may be a better way, but in all other places the creation method is used instead. So probably we should follow that pattern also.